test/screenshot/screenshot.js

```diff
@@ -1,60 +1,220 @@
+import {createHash} from 'crypto';
+import {readFile, writeFile} from 'fs';
+import {promisify} from 'util';
 import puppeteer from 'puppeteer';
 import compareImages from 'resemblejs/compareImages';
-import {promisify} from 'util';
-import {readFile, writeFile} from 'fs';
 import {assert} from 'chai';
+import Storage from '@google-cloud/storage';
 
 import comparisonOptions from './screenshot-comparison-options';
 
 const readFilePromise = promisify(readFile);
 const writeFilePromise = promisify(writeFile);
 
+const SERVICE_ACCOUNT_KEY = process.env.SERVICE_ACCOUNT_KEY;
+const BRANCH_NAME = process.env.BRANCH_NAME;
+const COMMIT_HASH = process.env.COMMIT_HASH;
+const GOLDENS_FILE_PATH = './test/screenshot/golden.json';
+const BUCKET_NAME = 'screenshot-uploads';
+const DEFAULT_METADATA = {
+  commit: COMMIT_HASH,
+  branch: BRANCH_NAME,
+};
+
+const GOLDEN = 'golden';
+const SNAPSHOT = 'snapshot';
+const DIFF = 'diff';
+
+let credentials;
+try {
+  credentials = JSON.parse(SERVICE_ACCOUNT_KEY);
+} catch (err) {
+  console.error('Service account key could not be parsed.');
+  process.exit(1);
+}
+
+const storage = new Storage({
+  credentials,
+});
+
+const bucket = storage.bucket(BUCKET_NAME);
+
 export default class Screenshot {
+  /**
+   * @param {string} urlPath The URL path to test
+   */
   constructor(urlPath) {
+    /** @private {string} */
     this.urlPath_ = urlPath;
-    this.imagePath_ = `${urlPath}.golden.png`;
-    this.snapshotImagePath_ = `${urlPath}.snapshot.png`;
-    this.diffPath_ = `${urlPath}.diff.png`;
     // TODO allow clients to specify capture-chrome options, like viewport size
   }
 
+  /**
+   * Returns the golden ID
+   * @return {string|undefined}
+   * @private
+   */
+  async getGoldenID_() {
+    const goldenFile = await readFilePromise(GOLDENS_FILE_PATH);
+    const goldenJSON = JSON.parse(goldenFile);
+    return goldenJSON[this.urlPath_];
+  }
+
+  /**
+   * Saves the golden ID
+   * @param {string} goldenId The hash of the golden image
+   * @private
+   */
+  async saveGoldenID_(goldenId) {
+    const goldenFile = await readFilePromise(GOLDENS_FILE_PATH);
+    const goldenJSON = JSON.parse(goldenFile);
+    goldenJSON[this.urlPath_] = goldenId;
+    await writeFilePromise(GOLDENS_FILE_PATH, JSON.stringify(goldenJSON, null, '  '));
+  }
+
+  /**
+   * Takes a screenshot of the URL
+   * @return {!Buffer}
+   * @private
+   */
+  async createScreenshotTask_() {
+    const browser = await puppeteer.launch();
+    const page = await browser.newPage();
+    await page.goto(`http://localhost:8080/${this.urlPath_}`);
+    const imageBuffer = await page.screenshot();
+    await browser.close();
+    return imageBuffer;
+  }
+
+  /**
+   * Generates a unique hash from an image's contents
+   * @param {!Buffer} imageBuffer The image buffer to hash
+   * @return {string}
+   * @private
+   */
+  generateImageID_(imageBuffer) {
+    return createHash('sha256').update(imageBuffer).digest('hex');
```

Comment:
Why not sha512?

Changing your hashing algorithm in the future is painful (just ask git or npm), so we should pick the "right" one from the start.

Reply:
No reason, I just chose the first one that came to mind